futures-util/src/future/shared.rs

```diff
@@ -201,38 +220,53 @@ where
             match res {
                 Poll::Pending => {
                     // Not ready, try to release the handle
-                    match this.inner.notifier.state.compare_and_swap(POLLING, IDLE, SeqCst) {
+                    match this
+                        .inner
+                        .notifier
+                        .state
+                        .compare_and_swap(POLLING, IDLE, SeqCst)
+                    {
                         POLLING => {
                             // Success
                             return Poll::Pending;
                         }
                         REPOLL => {
                             // Gotta poll again!
-                            let prev = this.inner.notifier.state.swap(POLLING, SeqCst);
+                            let prev =
+                                this.inner.notifier.state.swap(POLLING, SeqCst);
                             assert_eq!(prev, REPOLL);
                         }
                         _ => unreachable!(),
                     }
                 }
-                Poll::Ready(output) => {
-                    unsafe {
-                        *this.inner.future_or_output.get() =
-                            FutureOrOutput::Output(output.clone());
-                    }
+                Poll::Ready(output) => break output,
+            }
+        };
 
-                    // Complete the future
-                    let mut lock = this.inner.notifier.wakers.lock().unwrap();
-                    this.inner.notifier.state.store(COMPLETE, SeqCst);
-                    let wakers = &mut lock.take().unwrap();
-                    for (_key, opt_waker) in wakers {
-                        if let Some(waker) = opt_waker.take() {
-                            waker.wake();
-                        }
-                    }
-                    return Poll::Ready(output);
-                }
+        if let Some(_) = Arc::get_mut(&mut this.inner) {
```

Comment:
Nit: clippy is angry about this and wants you to write `if Arc::get_mut(&mut this.inner).is_some()` instead.